### PR TITLE
support 3rd-party backend

### PR DESCRIPTION
**Pull request by qiongerfei**
register for example:

```python
# register DEVICE_TYPE
c10::register_privateuse1_backend("npu")
torch.utils.rename_privateuse1_backend("npu")
# register DEVICE_MODULE
torch._register_device_module('npu', torch_npu.npu)
```

ussage:

"torch.cuda" should be replaced by "DEVICE_MODULE"
".cuda()" should be replaced by ".to(device=DEVICE_TYPE)"

### CODE REVIEW

**Comment by qiongerfei** [comment]
Here is another way to add third-party backend. Is it also acceptable in TorchTitan? https://github.com/pytorch/torchtune/pull/1826/files

**Comment by qiongerfei** [comment]
Could you please give me some advice for this PR? Thanks. @tianyu-l @fegin @wanchaol

File: torchtitan/metrics.py [hunk 1]
```diff
@@ -52,25 +53,25 @@ def _to_pct(self, memory):
...
-        num_ooms = cuda_info["num_ooms"]
+        num_retries = mem_info["num_alloc_retries"]
+        num_ooms = mem_info["num_ooms"]
 
         if num_retries > 0:
-            logger.warning(f"{num_retries} CUDA memory allocation retries.")
+            logger.warning(f"{num_retries} {DEVICE_TYPE} memory allocation retries.")
         if num_ooms > 0:
-            logger.warning(f"{num_ooms} CUDA OOM errors thrown.")
+            logger.warning(f"{num_ooms} {DEVICE_TYPE} OOM errors thrown.")
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: nit: `DEVICE_TYPE` -> `DEVICE_TYPE.upper` to keep them literally the same

File: train.py [hunk 2]
```diff
@@ -61,16 +61,16 @@ def main(job_config: JobConfig):
         world_size=world_size,
         enable_loss_parallel=job_config.training.enable_loss_parallel,
     )
-    device = torch.device(f"cuda:{int(os.environ['LOCAL_RANK'])}")
-    torch.cuda.set_device(device)
+    device = torch.device(f"{DEVICE_TYPE}:{int(os.environ['LOCAL_RANK'])}")
+    DEVICE_MODULE.set_device(device)
     utils.init_distributed(job_config)
     # initialize GPU memory monitor and get peak flops for MFU calculation
```
Comment (tianyu-l) [inline train.py]: pls modify comments

File: torchtitan/metrics.py [hunk 3]
```diff
@@ -11,13 +11,14 @@
 
 import torch
 from torch.utils.tensorboard import SummaryWriter
+from torchtitan import DEVICE_TYPE, DEVICE_MODULE
 from torchtitan.config_manager import JobConfig
 from torchtitan.logging import logger
 from torchtitan.parallelisms import ParallelDims
 
 # named tuple for passing GPU memory stats for logging
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: pls modify comments as well

File: torchtitan/metrics.py [hunk 4]
```diff
@@ -11,13 +11,14 @@
...
-GPUMemStats = namedtuple(
-    "GPUMemStats",
+MemStats = namedtuple(
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: for consistency, let's name it to `DeviceMemStats`, and also similar for other occurrences of `mem_monitor`, `build_memory_monitor`

**Review by tianyu-l — COMMENTED** [review]
looks good, thank you for contributing!
can you please address minor comments?

File: torchtitan/utils.py [hunk 5]
```diff
@@ -16,16 +16,17 @@
...
 def dist_max(x: Union[int, float], mesh: DeviceMesh) -> float:
-    tensor = torch.tensor(x).cuda()
+    tensor = torch.tensor(x).to(DEVICE_TYPE)
     return funcol.all_reduce(tensor, reduceOp=c10d.ReduceOp.MAX.name, group=mesh).item()
 
 
 def dist_mean(x: Union[int, float], mesh: DeviceMesh) -> float:
-    tensor = torch.tensor(x).cuda()
+    tensor = torch.tensor(x).to(DEVICE_TYPE)
     return funcol.all_reduce(tensor, reduceOp=c10d.ReduceOp.AVG.name, group=mesh).item()
```
Comment (kwen2501) [inline torchtitan/utils.py]: Is there a way to create tensor on a concrete device with index? Calling `.to(DEVICE_TYPE)` can lead to tensor creation on cuda:0 unintentionally. 
Can be in a separate PR as this seems to be an existing issue.
cc @tianyu-l

File: torchtitan/__init__.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+from torch._utils import _get_available_device_type, _get_device_module
+
+DEVICE_TYPE = _get_available_device_type()
+DEVICE_MODULE = _get_device_module(DEVICE_TYPE)
```
Comment (kwen2501) [inline torchtitan/__init__.py]: Can we keep these in utils as helper functions? 
Having in global scope may lead to them being executed when torchtitan is just imported. Just want to be cautious about potential side effects of these calls (such as creating a device context).

Comment (zqwenn) [reply torchtitan/__init__.py]: Yes, the helper function is added as below in util, pls  review it again, thanks!   @kwen2501 
``` python
from torch._utils import _get_available_device_type, _get_device_module
def get_device_info():
    device_type = _get_available_device_type()
    if device_type == None:
        device_type = "cuda"  # default device_type: cuda
    device_module = _get_device_module(device_type)  # default device_module:torch.cuda
    return device_type, device_module
```

**Comment by qiongerfei** [comment]
Hi @tianyu-l, I’ve made the code changes based on your suggestions; could you please review it again? Thanks☺️
Additionally:
1. modified comments "GPU memory monitor “ to ”device memory monitor"
2. rename "GPUMemStats" as "DeviceMemStats"

File: train.py [hunk 7]
```diff
@@ -144,9 +145,9 @@ def loss_fn(pred, labels):
         buffer_device = None
     elif job_config.training.enable_cpu_offload:
         init_device = "cpu"
-        buffer_device = "cuda"
+        buffer_device = f"{device_type}"
```
Comment (tianyu-l) [inline train.py]: can we just do
```suggestion
        buffer_device = device_type
```

File: train.py [hunk 8]
```diff
@@ -144,9 +145,9 @@ def loss_fn(pred, labels):
...
     else:
-        init_device = "cuda"
+        init_device = f"{device_type}"
```
Comment (tianyu-l) [inline train.py]: same here

**Review by tianyu-l — APPROVED** [review]
Looks good to me. Please address final inline comments.

File: torchtitan/utils.py [hunk 9]
```diff
@@ -16,19 +16,31 @@
 import torch
 import torch.distributed._functional_collectives as funcol
 import torch.distributed.distributed_c10d as c10d
+from torch._utils import _get_available_device_type, _get_device_module
```
Comment (kwen2501) [inline torchtitan/utils.py]: nit: sorry for missing to mention this in previous reviews, but it seems torch have formalized these APIs a bit.
https://pytorch.org/docs/stable/generated/torch.get_device_module.html

File: torchtitan/utils.py [hunk 10]
```diff
@@ -16,19 +16,31 @@
...
 from torch.distributed.device_mesh import DeviceMesh
-from torchtitan import DEVICE_TYPE, DEVICE_MODULE
+from torch.distributed.tensor import DTensor
 from torchtitan.logging import logger
 
 
+def get_device_info():
+    device_type = _get_available_device_type()
+    if device_type is None:
+        device_type = "cuda"  # default device_type: cuda
```
Comment (kwen2501) [inline torchtitan/utils.py:29]: nit: I am not sure if this fallback helps? Maybe just raise an error here?

**Review by kwen2501 — APPROVED** [review]
LGTM. Left some minor comments. Would be nice if you could accommodate them before merge. Thanks!

File: torchtitan/utils.py (hunk 9 above)
Comment (tianyu-l) [reply torchtitan/utils.py]: I can clean it up later. But anyways, it seems only one of the imports can be replaced with public API.
